### ISSUE TITLE
Enable analytic Jacobians for daeMode

### ISSUE THREAD

**Issue by OpenModelica-TracImporter** [OpenModelica/OpenModelica#6346]
By running this simple model

```julia
model M
  Real x(start = 1, fixed = true),y;
  
equation
  der(x) = -y;
  y + 0.1*sin(y) = x;
  annotation(__OpenModelica_commandLineOptions="--daeMode --generateSymbolicJacobian",
             __OpenModelica_simulationFlags(jacobian="coloredSymbolical",lv="LOG_STATS"));
end M;
```

I understand that analytic jacobians are not generated in DAE mode, since there is no trace of their computation in the C code and in the declarative debugger.

This is not very clear from the documentation of the `--generateSymbolicJacobian` [​compiler flag documentation](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-generatesymbolicjacobian) , which makes no explicit reference to either odeMode nor daeMode, but is clear from the description of [​postOptModules](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-postoptmodules), which states

symbolicJacobian (Detects the sparse pattern of the ODE system and calculates also the symbolic Jacobian if flag `--generateSymbolicJacobian` is enabled.)

In fact, [​this line of code](https://github.com/OpenModelica/OpenModelica/blob/c8233fa62ad0c3e01be24d6e4f97134c02204a10/OMCompiler/SimulationRuntime/c/simulation/solver/ida_solver.c#L345) should print out an appropriate error message when I try to turn on analytic jacobians in DAE mode, as in the above message. Unfortunately, if I do it in OMEdit, the simulation output is not passed to the simulation output window, so I can't see that (why?).

For the sake of high-speed power grid simulation, computing the jacobian of the DAE in daeMode would be very useful, since most of its nonzero coefficients are parameters and could be computed only once at initialization.

----

Migrated-From: https://trac.openmodelica.org/OpenModelica/ticket/6346

**Comment by casella** [comment]
Main goal: implement the symbolic computation of Jacobians for daeMode integration, to the extent that is necessary to support simulation of electro-mechanical models of AC power systems, e.g. including operator records for complex numbers, but (possibly) excluding handling of high-index problems (such models are described by index-1 DAEs) and the use of external functions.

Parallelization could also be exploited if time allows.

It would be good to exploit the constant part of the Jacobian – already done in the causalized backend, [paper](https://dx.doi.org/10.3384/ECP20169159) published at the American Modelica Conference 2020. @AnHeuermann worked on that, not finished, not working in all cases - there is one function that computes all the constant stuff and pulls it out of the online computations. Only if time allows.

**Comment by AnHeuermann** [comment]
> It would be good to exploit the constant part of the Jacobian – already done in the causalized backend, [paper](https://dx.doi.org/10.3384/ECP20169159) published at the American Modelica Conference 2020. @AnHeuermann worked on that, not finished, not working in all cases - there is one function that computes all the constant stuff and pulls it out of the online computations. Only if time allows.

I guess we need to enable analytic Jacobians and verify that they are working before starting optimizing that.This can be a future issue, so I'm removing my assignment for now.

**Comment by casella** [comment]
Analytic Jacobians for daeMode were implemented last summer by @kabdelhak. You can test all the possible combinations (odeMode, daeMode, with or without analytic jacobians) with the following test package:
```modelica
package TestDAEJacobian
  model NumericalODE
    parameter Real a = 0.001;
    parameter Real x0 = 0.5;
    Real x(start = 0, fixed = true);
  equation
    der(x) = 0.01 + exp(-((x - x0)^2/a));
    annotation(
      __OpenModelica_commandLineOptions = "",
      experiment(StartTime = 0, StopTime = 100, Tolerance = 1e-06, Interval = 0.02));
  end NumericalODE;

  model AnalyticODE
    extends NumericalODE;
    annotation(
      __OpenModelica_commandLineOptions = "--generateSymbolicJacobian",
      experiment(StartTime = 0, StopTime = 100, Tolerance = 1e-06, Interval = 0.02));
  end AnalyticODE;

  model NumericalDAE
    extends NumericalODE;
    annotation(
      __OpenModelica_commandLineOptions = "--daeMode",
      experiment(StartTime = 0, StopTime = 10, Tolerance = 1e-06, Interval = 0.02));
  end NumericalDAE;

  model AnalyticDAE
    extends NumericalODE;
    annotation(
      __OpenModelica_commandLineOptions = "--daeMode --generateSymbolicJacobian",
      experiment(StartTime = 0, StopTime = 10, Tolerance = 1e-06, Interval = 0.02));
  end AnalyticDAE;
end TestDAEJacobian;
```
When compiling the AnalyticDAE model it is possible to check with the Declarative Debugger that equations are generated to compute the terms of the Jacobian.

**Comment by casella** [comment]
The attached [TestAnalyticJacobiansGrids.mo.txt](https://github.com/OpenModelica/OpenModelica/files/10844821/TestAnalyticJacobiansGrids.mo.txt) package allows to compare the performance of a test contained in ScalableTestGrids having about 12,000 eqations, with and without analytic Jacobian generation.

These are the runtime statistics of `TestNumerical` on my PC:
```
timer
    0.474326s          reading init.xml
     0.12517s          reading info.xml
   0.0123454s [  3.7%] pre-initialization
   0.0175196s [  5.3%] initialization
   0.0002908s [  0.1%] steps
    0.106368s [ 32.0%] solver (excl. callbacks)
   0.0528147s [ 15.9%] creating output-file
   0.0076938s [  2.3%] event-handling
   0.0264634s [  8.0%] overhead
     0.10894s [ 32.8%] simulation
    0.332435s [100.0%] total

solver: ida
  297 steps taken
  585 calls of functionODE
   50 evaluations of jacobian
    4 error test failures
    0 convergence test failures
0.0393354s time of jacobian evaluation
```
while these are the runtime statistics of `TestAnalytic`, which sets `--generateSymbolicJacobian`, on the same computer:
```
timer
    0.394734s          reading init.xml
    0.118118s          reading info.xml
   0.0116859s [  4.6%] pre-initialization
   0.0144712s [  5.7%] initialization
   0.0003919s [  0.2%] steps
   0.0739738s [ 28.9%] solver (excl. callbacks)
    0.037786s [ 14.8%] creating output-file
   0.0051618s [  2.0%] event-handling
   0.0213197s [  8.3%] overhead
   0.0907449s [ 35.5%] simulation
    0.255535s [100.0%] total

solver: ida
  297 steps taken
  585 calls of functionODE
   50 evaluations of jacobian
    4 error test failures
    0 convergence test failures
0.012776s time of jacobian evaluation
```
The effect is clearly favourable, in terms of reduction of Jacobian evaluation time and consequently on the reduction of the simulation time. Also the solver performance improves, maybe because the more precise Jacobian allows to reduce the number of iterations to get convergence of the implicit BDF formula used by the IDA solver. Unfortunately I have no way to confirm that directly.

I will add `--generateSymbolicJacobians` to the CI tests of PowerGrids and ScalableTestGrids, to get faster simulation in those cases as well.
